File: rust/src/leetcode/Q416_partition_equal_subset_sum.rs
/*
 * @lc app=leetcode id=416 lang=rust
 *
 * [416] Partition Equal Subset Sum
 */

// @lc code=start
impl Solution {
    pub fn can_partition(nums: Vec<i32>) -> bool {
        let sum: i32 = nums.iter().sum::<i32>();

        if sum % 2 != 0 {
            return false;
        }

        let target: i32 = sum / 2;
        let mut dp: Vec<bool> = vec![false; target as usize + 1];
        dp[0] = true;

        for &num in &nums {
            for i in (num as usize..=target as usize).rev() {
                dp[i] = dp[i] || dp[i - num as usize];
            }
        }

        dp[dp.len() - 1]
    }
}
// @lc code=end

File: python/Q416_partition_equal_subset_sum.py
#
# @lc app=leetcode id=416 lang=python3
#
# [416] Partition Equal Subset Sum
#


# @lc code=start
class Solution:
    def canPartition(self, nums: list[int]) -> bool:
        target: int = sum(nums)
        if target % 2 == 1:
            return False

        target = target // 2

        dp: list[list[bool]] = [[False] * (target + 1) for _ in range(len(nums) + 1)]
        for i in range(len(nums) + 1):
            dp[i][0] = True

        for i in range(1, len(nums) + 1):
            element: int = nums[i - 1]

            for j in range(1, target + 1):
                dp[i][j] |= dp[i - 1][j]

                if j >= element:
                    dp[i][j] |= dp[i - 1][j - element]

        return dp[-1][-1]


# @lc code=end
